app/gui/src/controller/searcher/input.rs

```diff
@@ -0,0 +1,401 @@
+
+/// The description of the results of inserting suggestion.
+#[derive(Clone, Debug)]
+pub struct InsertedSuggestion {
+    /// An entire input after inserting the suggestion.
+    pub new_input:     String,
+    /// The replaced range in the old input.
+    pub replaced:      text::Range<text::Byte>,
+    /// The range of the inserted code in the new input. It does not contain any additional spaces
+    /// (in contrary to [`inserted_text`] field.
+    pub inserted_code: text::Range<text::Byte>,
+    /// The range of the entire inserted text in the new input. It contains code and all additional
+    /// spaces.
+    pub inserted_text: text::Range<text::Byte>,
+}
+
+
+impl InsertedSuggestion {
+    /// The text change resulting from this insertion.
+    pub fn input_change(&self) -> text::Change<text::Byte, String> {
+        let to_insert = self.new_input[self.inserted_text].to_owned();
+        text::Change { range: self.replaced, text: to_insert }
+    }
+}
+
+impl Input {
+    /// Return an information about input change after inserting given suggestion.
+    pub fn after_inserting_suggestion(
+        &self,
+        suggestion: &action::Suggestion,
+        has_this: bool,
+    ) -> FallibleResult<InsertedSuggestion> {
+        let replaced = self.replaced_range();
+        let generate_this = self.context().is_none() && !has_this;
+        let code_to_insert = suggestion.code_to_insert(generate_this);
+        debug!("Code to insert: \"{code_to_insert}\"");
+        let end_of_inserted_code = replaced.start + text::Bytes(code_to_insert.len());
+        let end_of_inserted_text = end_of_inserted_code + text::Bytes(1);
+        let mut new_input = self.ast.to_string();
+        let raw_range = replaced.start.value..replaced.end.value;
+        let range_start = raw_range.start;
+        let range_end = raw_range.end;
+        Self::ensure_on_char_boundary(&new_input, range_start)?;
+        Self::ensure_on_char_boundary(&new_input, range_end)?;
+        new_input.replace_range(raw_range, &code_to_insert);
+        new_input.insert(end_of_inserted_code.value, ' ');
+        Ok(InsertedSuggestion {
+            new_input,
+            replaced,
+            inserted_code: (replaced.start..end_of_inserted_code).into(),
+            inserted_text: (replaced.start..end_of_inserted_text).into(),
+        })
+    }
+
+    /// Return `true` if the `index` is on a char boundary inside `string` and can be used as a
+    /// range boundary. Otherwise, return `false` and report an error.
+    fn ensure_on_char_boundary(string: &str, index: usize) -> FallibleResult<()> {
+        if !string.is_char_boundary(index) {
+            Err(NotACharBoundary { index, string: string.into() }.into())
+        } else {
+            Ok(())
+        }
+    }
+}
+
+#[derive(Clone, Debug, Fail)]
+#[allow(missing_docs)]
+#[fail(display = "Not a char boundary: index {} at '{}'.", index, string)]
+pub struct NotACharBoundary {
+    index:  usize,
+    string: String,
+}
```

Comment:
Let's put it in `Errors` section somewhere at the top.

Reply:
Done